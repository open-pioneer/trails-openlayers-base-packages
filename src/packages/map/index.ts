// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
export * from "./api";
