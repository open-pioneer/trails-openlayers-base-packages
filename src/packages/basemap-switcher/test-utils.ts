// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
