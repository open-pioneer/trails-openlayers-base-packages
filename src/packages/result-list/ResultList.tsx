// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { Box } from "@open-pioneer/chakra-integration";
import { BaseFeature } from "@open-pioneer/map";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { useIntl } from "open-pioneer:react-hooks";
import { FC, ReactNode, RefObject, useEffect, useMemo, useRef, useState } from "react";
import { DataTable } from "./DataTable/DataTable";
import { createColumns } from "./DataTable/createColumns";
import { FormatNumberOptions } from "@formatjs/intl";

/**
 * Configures a column in the result list component.
 *
 * A column typically renders a property from the underlying feature.
 */
export interface ResultColumn {
    /**
     * Use this option to define an explicit column id.
     * This can be helpful to track your column when it moves in the table (for example, the sort order can be maintained).
     *
     * If this is not defined, {@link propertyName} will serve as a fallback.
     * If that is also not defined, the column index will be used instead.
     *
     * It is recommended to specify an id, if no {@link propertyName} has been set, because the
     * column index fallback is not advised.
     */
    id?: string;

    /**
     * The display name of this column.
     *
     * If no `displayName` has been configured, {@link propertyName} will serve as a fallback value.
     * If `propertyName` is also undefined, no column header will be rendered at all.
     */
    displayName?: string;

    /**
     * The width of this column, in pixels.
     */
    width?: number;

    /**
     * The property name to render.
     *
     * The value is expected to be available as `feature.properties[propertyName]`.
     *
     * See also {@link getPropertyValue}.
     */
    propertyName?: string;

    /**
     * Define this function to return a custom property value for this column.
     *
     * This can be used to create derived columns (by combining multiple properties into one value)
     * or to create columns for property that don't exist directly on the feature.
     *
     * The return value of this function will be rendered by the table.
     */
    getPropertyValue?: (feature: BaseFeature) => unknown;

    /** Custom render function to render a table cell in this column. */
    renderCell?: (context: RenderCellContext) => ReactNode;
}

/**
 * The arguments passed to {@link ResultColumn.renderCell | renderCell}.
 */
export interface RenderCellContext {
    /**
     * The feature in this row.
     */
    feature: BaseFeature;

    /**
     * The value of this column.
     * May be undefined if neither `propertyName` nor `getPropertyValue` was specified on the column.
     */
    value: unknown;
}

/**
 * To specify the format of cell values if they are of number or date type.
 */
export interface FormatOptions {
    /**
     * To specify the format of number type values
     */
    numberOptions?: FormatNumberOptions;
    /**
     *  To specify the format of date type values
     */
    dateOptions?: Intl.DateTimeFormatOptions;
}

/**
 * Configures the result list's content.
 */
export interface ResultListInput {
    /**
     * Configures the columns shown by the result list.
     */
    columns: ResultColumn[];

    /**
     * The data shown by the result list component.
     * Every feature will be rendered as an individual row.
     */
    data: BaseFeature[];

    /**
     * Optional formatOptions to specify the `numberOptions` for number type values and
     * `dateOptions` to specify the format of date type values
     */
    formatOptions?: FormatOptions;
}

/**
 * Emitted when the selection within the ResultList changes.
 */
export interface ResultListSelectionChangeEvent {
    features: BaseFeature[];
    getFeatureIds: () => (number | string)[];
}

/**
 * Properties supported by the {@link ResultList} component.
 */
export interface ResultListProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId?: string;

    /**
     * Describes the data rendered by the component.
     */
    input: ResultListInput;

    /**
     * This handler is called whenever the user has changed the selected features in the result-list
     */
    onSelectionChange?: (event: ResultListSelectionChangeEvent) => void;
}

/**
 * A component that displays a set of features as a list.
 */
export const ResultList: FC<ResultListProps> = (props) => {
    const { containerProps } = useCommonComponentProps("result-list", props);
    const intl = useIntl();
    const {
        input: { data, columns, formatOptions },
        onSelectionChange
    } = props;
    if (columns.length === 0) {
        throw Error("No columns were defined. The result list cannot be displayed.");
    }

    const containerRef = useRef<HTMLDivElement>(null);
    const tableWidth = useTableWidth(containerRef);
    const dataTableColumns = useMemo(
        () =>
            createColumns({
                columns: columns,
                intl: intl,
                tableWidth: tableWidth,
                formatOptions: formatOptions
            }),
        [columns, intl, tableWidth, formatOptions]
    );

    return (
        <Box {...containerProps} height="100%" overflowY="auto" ref={containerRef}>
            <DataTable
                columns={dataTableColumns}
                data={data}
                onSelectionChange={onSelectionChange}
            />
        </Box>
    );
};

function useTableWidth(tableRef: RefObject<HTMLDivElement> | null) {
    const [tableWidth, setTableWidth] = useState<number>();

    useEffect(() => {
        if (!tableRef?.current) return;
        const resizeObserver = new ResizeObserver((event) => {
            // Depending on the layout, you may need to swap inlineSize with blockSize
            // https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry/contentBoxSize
            const width = event[0]?.contentBoxSize[0]?.inlineSize;
            if (width != null) {
                setTableWidth(width);
            }
        });
        resizeObserver.observe(tableRef.current);
        return () => resizeObserver.disconnect();
    }, [tableRef]);
    return tableWidth;
}
