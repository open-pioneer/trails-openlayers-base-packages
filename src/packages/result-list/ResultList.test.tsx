// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { BaseFeature } from "@open-pioneer/map";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import { act, fireEvent, render, screen, waitFor } from "@testing-library/react";
import { SpyInstance, afterEach, beforeEach, expect, it, vi } from "vitest";
import { ResultColumn, ResultList, ResultListInput } from "./ResultList";
import { Point } from "ol/geom";

afterEach(() => {
    vi.restoreAllMocks();
});

let errorSpy!: SpyInstance;
beforeEach(() => {
    errorSpy = vi.spyOn(console, "error");
});

function doNothing() {}

it("expect result list to be created successfully", async () => {
    render(
        <PackageContextProvider>
            <ResultList
                input={{ data: dummyFeatureData, columns: dummyColumns }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { resultListDiv } = await waitForResultList();
    expect(resultListDiv).toMatchSnapshot();
});

it("expect result list column and row count to match data/metadata", async () => {
    render(
        <PackageContextProvider>
            <ResultList
                input={{ data: dummyFeatureData, columns: dummyColumns }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { allHeaderElements, allRows } = await waitForResultList();

    // +1 because of the added selection column
    expect(allHeaderElements.length).toEqual(dummyColumns.length + 1);
    expect(allRows.length).toEqual(dummyFeatureData.length);
});

it("expect empty data text to be shown", async () => {
    const emptyData: ResultListInput = {
        data: [],
        columns: dummyColumns
    };

    let error;
    try {
        render(
            <PackageContextProvider>
                <ResultList input={emptyData} mapId="foo" data-testid="result-list" />
            </PackageContextProvider>
        );
    } catch (e) {
        error = new Error("unexpected failure");
    }

    const { resultListDiv } = await waitForResultList();

    expect(error).not.toBeDefined();
    expect(resultListDiv.textContent).toEqual("noDataMessage");
    expect(resultListDiv).toMatchSnapshot();
});

it("expect empty metadata to throw error", async () => {
    errorSpy.mockImplementation(doNothing);

    const emptyMetadata: ResultListInput = {
        data: dummyFeatureData,
        columns: []
    };

    expect(() => {
        render(
            <PackageContextProvider>
                <ResultList input={emptyMetadata} mapId="foo" data-testid="result-list" />
            </PackageContextProvider>
        );
    }).toThrowErrorMatchingSnapshot();

    expect(errorSpy).toHaveBeenCalledOnce();
});

it("expect getPropertyValue to be used correctly", async () => {
    const getPropertyValueMock = vi.fn((_feature) => {
        return "virtual property";
    });
    const dummyFeatureData: BaseFeature[] = [
        {
            id: "1",
            properties: {
                "b": "123",
                "c": undefined
            },
            geometry: undefined
        }
    ];
    const columns: ResultColumn[] = [
        {
            propertyName: "properties.b",
            displayName: "Spalte B",
            width: 50,
            getPropertyValue: getPropertyValueMock
        }
    ];
    const resultListInput = {
        data: dummyFeatureData,
        columns: columns
    };

    render(
        <PackageContextProvider>
            <ResultList input={resultListInput} mapId="foo" data-testid="result-list" />
        </PackageContextProvider>
    );

    const { allRows } = await waitForResultList();

    expect(getPropertyValueMock).toHaveBeenCalled();
    expect(getPropertyValueMock).toHaveBeenCalledWith(dummyFeatureData[0]);
    expect(allRows.item(0).children[1]?.textContent).toEqual("virtual property");
});

it("expect changes of data and metadata to change full table", async () => {
    const renderResult = render(
        <PackageContextProvider>
            <ResultList
                input={{ data: dummyFeatureData, columns: dummyColumns }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { allHeaderElements, allRows } = await waitForResultList();

    // +1 because of the added selection column
    expect(allHeaderElements.length).toEqual(dummyColumns.length + 1);
    expect(allRows.length).toEqual(dummyFeatureData.length);

    renderResult.rerender(
        <PackageContextProvider>
            <ResultList
                input={{ data: dummyFeatureDataAlt, columns: dummyMetaDataAlt }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { allHeaderElements: allHeaderElementsAlt, allRows: allRowsAlt } =
        await waitForResultList();

    // Ensure dummydata is different
    expect(allHeaderElements.length).not.toEqual(allHeaderElementsAlt.length);
    expect(allRows.length).not.toEqual(allRowsAlt.length);
    // +1 because of the added selection column
    expect(allHeaderElementsAlt.length).toEqual(dummyMetaDataAlt.length + 1);
    expect(allRowsAlt.length).toEqual(dummyFeatureDataAlt.length);
});

it("expect selection column to be added", async () => {
    render(
        <PackageContextProvider>
            <ResultList
                input={{ data: dummyFeatureData, columns: dummyColumns }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { selectAllSelect, selectRowSelects } = await waitForResultList();
    expect(selectAllSelect).toBeDefined();
    expect(selectRowSelects).toBeDefined();
    expect(selectRowSelects.length).toEqual(dummyFeatureData.length);
});

it("expect all rows to be selected and deselected", async () => {
    render(
        <PackageContextProvider>
            <ResultList
                input={{ data: dummyFeatureData, columns: dummyColumns }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { selectAllSelect, selectRowSelects } = await waitForResultList();
    expect(selectAllSelect).toBeDefined();
    expect(selectRowSelects).toBeDefined();

    expect(selectAllSelect!.checked).toBeFalsy();
    selectRowSelects.forEach((checkbox) => expect(checkbox.checked).toBeFalsy());

    act(() => {
        fireEvent.click(selectAllSelect!);
    });

    expect(selectAllSelect!.checked).toBeTruthy();
    selectRowSelects.forEach((checkbox) => expect(checkbox.checked).toBeTruthy());

    act(() => {
        fireEvent.click(selectAllSelect!);
    });

    expect(selectAllSelect!.checked).toBeFalsy();
    selectRowSelects.forEach((checkbox) => expect(checkbox.checked).toBeFalsy());
});

it("expect result list display all data types except dates", async () => {
    render(
        <PackageContextProvider locale="de">
            <ResultList
                input={{
                    data: dummyFeatureData,
                    columns: dummyColumns,
                    formatOptions: {
                        numberOptions: { maximumFractionDigits: 3 }
                    }
                }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { allRows } = await waitForResultList();
    const firstRowCells = Array.from(allRows[0]!.querySelectorAll("td"));
    expect(firstRowCells).toHaveLength(6);

    const [selectCell, stringCell, integerCell, floatCell, trueCell] = firstRowCells;
    expect(selectCell!.innerHTML).includes("<input");
    expect(stringCell!.textContent).toBe("Test");
    expect(integerCell!.textContent).toBe("123");
    expect(floatCell!.textContent).toBe("4,567");
    expect(trueCell!.textContent).toBe("displayBoolean.true");

    const falseCell = allRows[1]?.querySelectorAll("td")[4];
    expect(falseCell!.textContent).toBe("displayBoolean.false");

    // Null / Undefined is rendered as an empty string
    const lastRowCells = Array.from(allRows[3]!.querySelectorAll("td"));
    expect(lastRowCells).toHaveLength(6);
    for (let i = 0; i < 6; ++i) {
        const cell = lastRowCells[i]!;
        expect(cell.textContent, "cell " + i).toBe("");
    }
});

it("expect result list display date in given format", async () => {
    const dateTimeFormatOptions: Intl.DateTimeFormatOptions = {
        dateStyle: "medium",
        timeStyle: "medium",
        timeZone: "UTC"
    };

    let dateFormatter = Intl.DateTimeFormat("de-DE", dateTimeFormatOptions);
    const resultListComp = (
        <ResultList
            input={{
                data: dummyDateFeatureData,
                columns: dummyDateColumns,
                formatOptions: {
                    numberOptions: { maximumFractionDigits: 3 },
                    dateOptions: dateTimeFormatOptions
                }
            }}
            mapId="foo"
            data-testid="result-list"
        />
    );

    const renderResult = render(
        <PackageContextProvider locale="de">{resultListComp}</PackageContextProvider>
    );

    const { allRows } = await waitForResultList();
    const firstRowCells = Array.from(allRows[0]!.querySelectorAll("td"));
    const [_, dateCell] = firstRowCells;
    expect(dateCell!.textContent).toBe(dateFormatter.format(new Date("2020-05-12T23:50:21.817Z")));

    renderResult.rerender(
        <PackageContextProvider locale="en">{resultListComp}</PackageContextProvider>
    );
    await waitForResultList(); // TODO: Workaround to hide react warning due to useEffect (use disableReactWarning helper after printing merge)

    dateFormatter = Intl.DateTimeFormat("en-US", dateTimeFormatOptions);
    expect(dateCell!.textContent).toBe(dateFormatter.format(new Date("2020-05-12T23:50:21.817Z")));
});

it("expect render function to be applied", async () => {
    render(
        <PackageContextProvider locale="de">
            <ResultList
                input={{
                    data: dummyDateFeatureData,
                    columns: dummyColumnsWithRenderFunc
                }}
                mapId="foo"
                data-testid="result-list"
            />
        </PackageContextProvider>
    );

    const { allRows } = await waitForResultList();
    const firstRowCells = Array.from(allRows[0]!.querySelectorAll("td"));
    const [_, dateCell] = firstRowCells;
    expect(dateCell!.textContent).toMatchSnapshot();
});

async function waitForResultList() {
    return await waitFor(async () => {
        const resultListDiv: HTMLDivElement | null =
            await screen.findByTestId<HTMLDivElement>("result-list");
        if (!resultListDiv) {
            throw new Error("Result list not rendered");
        }

        const allHeaderElements =
            resultListDiv.querySelectorAll<HTMLTableHeaderCellElement>("thead tr th");

        const allRows = resultListDiv.querySelectorAll<HTMLElement>("tbody tr");

        const selectAllSelect = resultListDiv.querySelector<HTMLInputElement>(
            ".result-list-select-all-checkbox input"
        );

        const selectRowSelects = resultListDiv.querySelectorAll<HTMLInputElement>(
            ".result-list-select-row-checkbox input"
        );

        return {
            resultListDiv,
            allHeaderElements,
            allRows,
            selectAllSelect,
            selectRowSelects
        };
    });
}

const DATE_FORMAT = Intl.DateTimeFormat("de-DE", {
    dateStyle: "full",
    timeStyle: "full",
    timeZone: "UTC"
});

// Stable date format for tests.
function formatDate(date: Date) {
    return DATE_FORMAT.format(date);
}

const dummyDateFeatureData: BaseFeature[] = [
    {
        id: "1",
        properties: {
            "a": new Date("2020-05-12T23:50:21.817Z")
        },
        geometry: undefined
    }
];

const dummyDateColumns: ResultColumn[] = [
    {
        propertyName: "a",
        displayName: "Spalte A",
        width: 100
    }
];

const dummyColumnsWithRenderFunc: ResultColumn[] = [
    {
        propertyName: "a",
        displayName: "Spalte A",
        width: 100,
        renderCell: ({ feature }) => (
            <div className="renderTest">{`This item has the following ID: ${feature.id}`}</div>
        )
    }
];

const dummyFeatureData: BaseFeature[] = [
    {
        id: "1",
        properties: {
            "a": "Test",
            "b": 123,
            "c": 4.5671365,
            "d": true,
            "e": formatDate(new Date("2020-05-12T23:50:21.817Z"))
        },
        geometry: new Point([404567.3, 5757788.32])
    },
    {
        id: "2",
        properties: {
            "a": "Test123",
            "b": 434,
            "c": 78.567,
            "d": false,
            "e": formatDate(new Date("2021-05-12T23:50:21.817Z"))
        },
        geometry: new Point([406510.87, 5758314.82])
    },
    {
        id: "3",
        properties: {
            "a": "Testabc",
            "b": 666,
            "c": 8.597,
            "d": true,
            "e": formatDate(new Date("2020-10-12T23:30:21.817Z"))
        },
        geometry: new Point([406590.87, 5758311.82])
    },
    {
        id: "4",
        properties: {
            "a": null,
            "b": undefined,
            "c": "",
            "d": undefined,
            "e": undefined
        },
        geometry: new Point([406590.87, 5758311.82])
    },
    {
        id: "5",
        properties: {
            "a": NaN,
            "b": NaN,
            "c": NaN,
            "d": NaN,
            "e": NaN
        },
        geometry: undefined
    }
];

const dummyColumns: ResultColumn[] = [
    {
        propertyName: "a",
        displayName: "Spalte A",
        width: 100
    },
    {
        propertyName: "b",
        displayName: "Spalte B",
        width: 50
    },
    {
        propertyName: "c",
        displayName: "Spalte C",
        width: 150
    },
    {
        propertyName: "d",
        displayName: "Spalte D",
        width: 75
    },
    {
        propertyName: "e",
        displayName: "Spalte E",
        width: 50
    }
];

const dummyFeatureDataAlt: BaseFeature[] = [
    {
        id: "1",
        properties: {
            "f": "Test 42",
            "g": undefined
        },
        geometry: undefined
    }
];

const dummyMetaDataAlt: ResultColumn[] = [
    {
        propertyName: "f",
        displayName: "Spalte F",
        width: 200
    },
    {
        propertyName: "g",
        displayName: "Spalte G",
        width: 300
    }
];
