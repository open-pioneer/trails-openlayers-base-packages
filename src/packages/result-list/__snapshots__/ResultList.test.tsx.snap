// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expect empty data text to be shown 1`] = `
<div
  class="result-list css-u718rw"
  data-testid="result-list"
  data-theme="light"
>
  <div
    class="result-list-no-data-message css-0"
    data-theme="light"
  >
    noDataMessage
  </div>
</div>
`;

exports[`expect empty metadata to throw error 1`] = `[Error: No columns were defined. The result list cannot be displayed.]`;

exports[`expect render function to be applied 1`] = `"This item has the following ID: 1"`;

exports[`expect result list to be created successfully 1`] = `
<div
  class="result-list css-u718rw"
  data-testid="result-list"
  data-theme="light"
>
  <table
    class="chakra-table result-list-table css-lkg3hs"
    data-theme="light"
    style="--header-selection-buttons-size: 70; --col-selection-buttons-size: 70; --header-result-list-col_a-size: 100; --col-result-list-col_a-size: 100; --header-result-list-col_b-size: 50; --col-result-list-col_b-size: 50; --header-result-list-col_c-size: 150; --col-result-list-col_c-size: 150; --header-result-list-col_d-size: 75; --col-result-list-col_d-size: 75; --header-result-list-col_e-size: 50; --col-result-list-col_e-size: 50;"
  >
    <thead
      class="css-0"
      data-theme="light"
    >
      <tr
        class="result-list-headers css-0"
        data-theme="light"
      >
        <th
          aria-sort="none"
          class="result-list-header css-zdxa4r"
          data-theme="light"
          style="width: 50px;"
          tabindex="0"
        >
          <div
            class="result-list-select-all-container css-1baulvz"
            data-theme="light"
          >
            <span
              class="css-0"
              data-theme="light"
            >
              <label
                class="chakra-checkbox result-list-select-all-checkbox css-1577qb8"
                data-theme="light"
              >
                <input
                  class="chakra-checkbox__input"
                  style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="checkbox"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="chakra-checkbox__control css-1ydjfm6"
                  data-theme="light"
                />
              </label>
            </span>
          </div>
          <span
            class="result-list-resizer css-0"
            data-theme="light"
          />
        </th>
        <th
          aria-sort="none"
          class="result-list-header css-y14pq5"
          data-theme="light"
          style=""
          tabindex="0"
        >
          Spalte A
          <span
            class="result-list-sort-indicator css-1q8b5yy"
            data-theme="light"
          />
          <span
            class="result-list-resizer css-0"
            data-theme="light"
          />
        </th>
        <th
          aria-sort="none"
          class="result-list-header css-y14pq5"
          data-theme="light"
          style=""
          tabindex="0"
        >
          Spalte B
          <span
            class="result-list-sort-indicator css-1q8b5yy"
            data-theme="light"
          />
          <span
            class="result-list-resizer css-0"
            data-theme="light"
          />
        </th>
        <th
          aria-sort="none"
          class="result-list-header css-y14pq5"
          data-theme="light"
          style=""
          tabindex="0"
        >
          Spalte C
          <span
            class="result-list-sort-indicator css-1q8b5yy"
            data-theme="light"
          />
          <span
            class="result-list-resizer css-0"
            data-theme="light"
          />
        </th>
        <th
          aria-sort="none"
          class="result-list-header css-y14pq5"
          data-theme="light"
          style=""
          tabindex="0"
        >
          Spalte D
          <span
            class="result-list-sort-indicator css-1q8b5yy"
            data-theme="light"
          />
          <span
            class="result-list-resizer css-0"
            data-theme="light"
          />
        </th>
        <th
          aria-sort="none"
          class="result-list-header css-y14pq5"
          data-theme="light"
          style=""
          tabindex="0"
        >
          Spalte E
          <span
            class="result-list-sort-indicator css-1q8b5yy"
            data-theme="light"
          />
          <span
            class="result-list-resizer css-0"
            data-theme="light"
          />
        </th>
      </tr>
    </thead>
    <tbody
      class="result-list-table-body css-0"
      data-theme="light"
    >
      <tr
        class="result-list-table-row css-0"
        data-theme="light"
      >
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          <div
            class="result-list-select-row-container css-1baulvz"
            data-theme="light"
          >
            <label
              class="chakra-checkbox result-list-select-row-checkbox css-1577qb8"
              data-theme="light"
            >
              <input
                aria-disabled="false"
                aria-label="ariaLabel.selectSingle"
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-1ydjfm6"
                data-theme="light"
              />
            </label>
          </div>
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          Test
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          123
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          4.567
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          displayBoolean.true
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          Dienstag, 12. Mai 2020 um 23:50:21 Koordinierte Weltzeit
        </td>
      </tr>
      <tr
        class="result-list-table-row css-0"
        data-theme="light"
      >
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          <div
            class="result-list-select-row-container css-1baulvz"
            data-theme="light"
          >
            <label
              class="chakra-checkbox result-list-select-row-checkbox css-1577qb8"
              data-theme="light"
            >
              <input
                aria-disabled="false"
                aria-label="ariaLabel.selectSingle"
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-1ydjfm6"
                data-theme="light"
              />
            </label>
          </div>
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          Test123
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          434
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          78.567
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          displayBoolean.false
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          Mittwoch, 12. Mai 2021 um 23:50:21 Koordinierte Weltzeit
        </td>
      </tr>
      <tr
        class="result-list-table-row css-0"
        data-theme="light"
      >
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          <div
            class="result-list-select-row-container css-1baulvz"
            data-theme="light"
          >
            <label
              class="chakra-checkbox result-list-select-row-checkbox css-1577qb8"
              data-theme="light"
            >
              <input
                aria-disabled="false"
                aria-label="ariaLabel.selectSingle"
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-1ydjfm6"
                data-theme="light"
              />
            </label>
          </div>
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          Testabc
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          666
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          8.597
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          displayBoolean.true
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          Montag, 12. Oktober 2020 um 23:30:21 Koordinierte Weltzeit
        </td>
      </tr>
      <tr
        class="result-list-table-row css-0"
        data-theme="light"
      >
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          <div
            class="result-list-select-row-container css-1baulvz"
            data-theme="light"
          >
            <label
              class="chakra-checkbox result-list-select-row-checkbox css-1577qb8"
              data-theme="light"
            >
              <input
                aria-disabled="false"
                aria-label="ariaLabel.selectSingle"
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-1ydjfm6"
                data-theme="light"
              />
            </label>
          </div>
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
      </tr>
      <tr
        class="result-list-table-row css-0"
        data-theme="light"
      >
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        >
          <div
            class="result-list-select-row-container css-1baulvz"
            data-theme="light"
          >
            <label
              class="chakra-checkbox result-list-select-row-checkbox css-1577qb8"
              data-theme="light"
            >
              <input
                aria-disabled="false"
                aria-label="ariaLabel.selectSingle"
                class="chakra-checkbox__input"
                style="border: 0px; clip: rect(0, 0, 0, 0); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-checkbox__control css-1ydjfm6"
                data-theme="light"
              />
            </label>
          </div>
        </td>
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
        <td
          class="result-list-table-row css-xumdn4"
          data-theme="light"
          style=""
        />
      </tr>
    </tbody>
  </table>
</div>
`;
